### PR TITLE
pkg/cmd/*: fix openshift operator clients and host-endpoint-controller

### PR DESCRIPTION
**Pull request by alaypatel07**
This fix:

- uses a dynamic informer for etcds.operator.openshift.io GVR
- a bug where etcd members were staying in Unknown state even 
after pod transitions in Running state
- creates a copy of hostendpoint before updating so there are no
 side effects of using the cache.

### CODE REVIEW

**Comment by alaypatel07** [comment]
infra `oauth2: cannot fetch token: Post https://oauth2.googleapis.com/token: dial tcp: i/o timeout`

/test e2e-gcp-upgrade

**Comment by alaypatel07** [comment]
/hold cancel

tested 4/5 successful manual bootstraps. 1 failure was unrelated to the code changed

File: pkg/cmd/staticsynccontroller/staticsynccontroller.go [hunk 1]
```diff
@@ -7,15 +7,18 @@ import (
 	"os"
 	"time"
 
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
```
Comment (hexfusion) [inline pkg/cmd/staticsynccontroller/staticsynccontroller.go]: why is this not with other k8s imports?

File: pkg/operator/bootstrapteardown/teardown.go [hunk 2]
```diff
@@ -2,14 +2,18 @@ package bootstrapteardown
 
 import (
 	"context"
+	"k8s.io/apimachinery/pkg/util/wait"
```
Comment (hexfusion) [inline pkg/operator/bootstrapteardown/teardown.go]: again something is making odd choices on where these go.

File: pkg/operator/bootstrapteardown/teardown.go [hunk 3]
```diff
@@ -20,13 +24,12 @@ import (
...
 func TearDownBootstrap(config *rest.Config,
-	clusterMemberShipController *clustermembercontroller.ClusterMemberController, operatorClient v1helpers.OperatorClient) error {
+	clusterMemberShipController *clustermembercontroller.ClusterMemberController, etcdClient etcdv1.EtcdInterface) error {
 	failing := configv1.ClusterStatusConditionType("Failing")
 	var lastError string
-	var err error
 
 	cc := configclient.NewForConfigOrDie(config)
-	_, err = clientwatch.UntilWithSync(
+	_, _ = clientwatch.UntilWithSync(
```
Comment (hexfusion) [inline pkg/operator/bootstrapteardown/teardown.go]: why dont we care about err?

File: pkg/operator/bootstrapteardown/teardown.go [hunk 4]
```diff
@@ -55,24 +58,32 @@ func TearDownBootstrap(config *rest.Config,
...
 
-	if err == nil {
 		klog.Infof("clusterversions is available, safe to remove bootstrap")
-		switch operatorSpec.ManagementState {
+		switch etcd.Spec.ManagementState {
 		case operatorv1.Managed:
 			if clusterMemberShipController.IsMember("etcd-bootstrap") {
-				return clusterMemberShipController.RemoveBootstrap()
+				//TODO: need to keep retry as long as we are sure bootstrap is not removed
+				err := clusterMemberShipController.RemoveBootstrap()
```
Comment (hexfusion) [inline pkg/operator/bootstrapteardown/teardown.go]: nit `if err := clusterMemberShipController.RemoveBootstrap(); err != nil {`

scope of error is in this block only

File: pkg/operator/hostetcdendpointcontroller/hostendpointcontroller.go [hunk 5]
```diff
@@ -143,13 +143,15 @@ func (h *HostEtcdEndpointController) sync() error {
 		return fmt.Errorf("unexpected length of host endpoint subset")
 	}
```
Comment (hexfusion) [inline pkg/operator/hostetcdendpointcontroller/hostendpointcontroller.go]: We need to revisit this method its very hard to follow what is going on and it seems like we should be able to do this simpler but for the sake of time it works for me.

Comment (alaypatel07) [reply pkg/operator/hostetcdendpointcontroller/hostendpointcontroller.go]: sure thanks, I would love to see how to simplify this. 

There are a bunch of corner cases I am taking care of so we don't break host-endpoint. It is critical to KAS calculating storage URLs so being extra careful

File: pkg/operator/bootstrapteardown/teardown.go (hunk 3 above)
Comment (alaypatel07) [reply pkg/operator/bootstrapteardown/teardown.go]: It was my understanding that because we never return error from the inner function we dont care about the error returned here.

Basically we retry until we don't succeed and ignore the error? do I need to change it? is it reasonable?

Comment (hexfusion) [reply pkg/operator/bootstrapteardown/teardown.go]: if there is some unknown issue I think it would be good to know could we print the error if klog level 4?

Comment (alaypatel07) [reply pkg/operator/bootstrapteardown/teardown.go]: sure

File: pkg/operator/bootstrapteardown/teardown.go [hunk 6]
```diff
@@ -55,24 +59,35 @@ func TearDownBootstrap(config *rest.Config,
 		},
 	)
 
-	operatorSpec, _, _, err := operatorClient.GetOperatorState()
 	if err != nil {
-		return err
+		klog.Errorf("error in watching clusterversions: %#v", err)
```
Comment (alaypatel07) [inline pkg/operator/bootstrapteardown/teardown.go]: not sure how to handle this @hexfusion 

Do we want to return here because we errored?

Comment (alaypatel07) [reply pkg/operator/bootstrapteardown/teardown.go]: nevermind, this is changing in subsequent PRs

File: pkg/operator/bootstrapteardown/teardown.go (hunk 4 above)
Comment (alaypatel07) [reply pkg/operator/bootstrapteardown/teardown.go]: done

File: pkg/operator/bootstrapteardown/teardown.go (hunk 2 above)
Comment (alaypatel07) [reply pkg/operator/bootstrapteardown/teardown.go]: fixed

File: pkg/cmd/staticsynccontroller/staticsynccontroller.go (hunk 1 above)
Comment (alaypatel07) [reply pkg/cmd/staticsynccontroller/staticsynccontroller.go]: fixed

**Comment by alaypatel07** [comment]
@hexfusion fixed all imports, PTAL

**Comment by alaypatel07** [comment]
looks like infra
```
level=error msg="Error: Provider produced inconsistent result after apply"
level=error
level=error msg="When applying changes to module.bootstrap.aws_s3_bucket.ignition, provider"
level=error msg="\"aws\" produced an unexpected new value for was present, but now absent."
level=error
```

/test e2e-aws

**Comment by alaypatel07** [comment]
```
Nov 26 19:20:23.376 E ns/openshift-sdn pod/sdn-controller-fjf6c node/ci-op--llklb-m-0.c.openshift-gce-devel-ci.internal container=sdn-controller container exited with code 2 (Error): I1126 18:41:53.654333       1 leaderelection.go:241] attempting to acquire leader lease  openshift-sdn/openshift-network-controller...\nE1126 18:47:08.624251       1 leaderelection.go:330] error retrieving resource lock openshift-sdn/openshift-network-controller: etcdserver: request timed out\nE1126 19:11:28.982427       1 leaderelection.go:330] error retrieving resource lock openshift-sdn/openshift-network-controller: configmaps "openshift-network-controller" is forbidden: User "system:serviceaccount:openshift-sdn:sdn-controller" cannot get resource "configmaps" in API group "" in the namespace "openshift-sdn": RBAC: [role.rbac.authorization.k8s.io "openshift-sdn-controller-leaderelection" not found, clusterrole.rbac.authorization.k8s.io "system:image-puller" not found]\n
```
hrm, weird! would like to see if this happens again

/test e2e-gcp-upgrade

**Comment by alaypatel07** [comment]
why?
```
error: could not run steps: step e2e-gcp-upgrade failed: could not wait for template instance to be ready: could not determine if template instance was ready: failed to create objects: object is being deleted: pods "e2e-gcp-upgrade" already exists
```
/test e2e-gcp-upgrade

**Comment by hexfusion** [comment]
```
level=warning msg="Found override for release image. Please be warned, this is not advised"
level=info msg="Consuming Install Config from target directory"
level=info msg="Creating infrastructure resources..."
level=info msg="Waiting up to 30m0s for the Kubernetes API at https://api.ci-op-6h4yy53c-e4498.origin-ci-int-gce.dev.openshift.com:6443..."
level=info msg="API v1.16.2 up"
level=info msg="Waiting up to 30m0s for bootstrapping to complete..."
level=info msg="Destroying the bootstrap resources..."
level=info msg="Waiting up to 30m0s for the cluster at https://api.ci-op-6h4yy53c-e4498.origin-ci-int-gce.dev.openshift.com:6443 to initialize..."
level=error msg="Cluster operator authentication Degraded is True with RouteHealthDegradedFailedGet: RouteHealthDegraded: failed to GET route: dial tcp 34.74.210.194:443: connect: connection refused"
level=info msg="Cluster operator authentication Progressing is Unknown with NoData: "
level=info msg="Cluster operator authentication Available is Unknown with NoData: "
level=info msg="Cluster operator console Progressing is True with SyncLoopRefreshProgressingInProgress: SyncLoopRefreshProgressing: Working toward version 0.0.1-2019-11-27-125951"
level=info msg="Cluster operator console Available is False with DeploymentAvailableFailedUpdate: DeploymentAvailable: 1 replicas ready at version 0.0.1-2019-11-27-125951"
level=info msg="Cluster operator insights Disabled is False with : "
level=info msg="Cluster operator support Disabled is True with NotAuthorized: Reporting was not allowed: your Red Hat account is not enabled for remote support or your token has expired"
level=fatal msg="failed to initialize the cluster: Working towards 0.0.1-2019-11-27-125951: 100% complete"
```
...

/retest
